### PR TITLE
✨ NEW: Writing workflows - "Basics" module

### LINKED ISSUE #339: 👌 IMPROVE: Module: Writing workflows - Basics
**General note:** Make any changes by opening a PR into the `tutorial-2021-intro-week` branch. I'm preparing a [wiki page with more information on MyST Markdown](https://github.com/aiidateam/aiida-tutorials/wiki/Moving-to-using-MyST-Markdown), but this is still a work in progress.

This module is part of the Workflows section, which currently has three modules:

* Running workflows
* **Writing workflows - Basics**
* Error handling

This module still needs quite a bit work in case we want to make it easier for participants to get started with writing workflows. @flavianojs maybe it's good that we first touch base on this before you start working on it? I'll already leave some notes below to get this started:

#### Code-agnostic modules

Since the content in this tutorial is already challenging enough as it is, I would stick to easy code-agnostic examples. This has the added benefit that in case we write a good code-agnostic tutorial on how to write workflows, it can also be re-used by other plugin developers.

#### Work functions

The tutorial now starts with work _functions_, as well as a short revision  of calculation functions. Maybe leaving in the work functions is fine to start simple, but of course the jump to work _chains_ is still massive. I think we can remove the calculation function part since it's already covered in the basics section.

#### Work chains

The biggest challenge will be to step by step explain the concepts for the work chain. I think this needs to really start with a _super_ basic example, almost minimal to have a working work chain. Step by step the participants can then add more to the work chain they are developing, building towards the `MultiplyAddWorkChain` _without_ the exit code. As described [here](https://documentation.divio.com/tutorials/), the tutorial should not explain too much, but rather have them learn in little steps by always expanding the work chain a little. I'd have to work on this a bit to be sure what these steps should be, but just off the top of my head:

1. How to write a define method, i.e.:
  * basics - using `super()`.
  * Defining an input via the `spec`. The `valid_type` argument.
  * Defining the outline via the `spec`. Writing the outline steps as methods of the class.
  * Defining an output via the `spec`.
  * Accessing an input inside the work chain.

This would be the minimal example (untested):

```python
class MultiplyAddWorkChain(WorkChain):
    """WorkChain to multiply two numbers and add a third, for testing and demonstration purposes."""

    @classmethod
    def define(cls, spec):
        """Specify inputs and outputs."""
        super().define(spec)
        spec.input("x", valid_type=Int)
        spec.outline(
            cls.result,
        )
        spec.output("result", valid_type=Int)

    def result(self):
        """Add the result to the outputs."""
        self.out("result", self.inputs.x)
``` 
2. How to use the context:
  * Pass variables between steps in the outline using the context.
  * Add another step to the _linear_ outline.
  * Note how the output has to be a node, but this is true because multiplying two `Int` nodes gives and `Int` node.

```python
class MultiplyAddWorkChain(WorkChain):
    """WorkChain to multiply two numbers and add a third, for testing and demonstration purposes."""

    @classmethod
    def define(cls, spec):
        """Specify inputs and outputs."""
        super().define(spec)
        spec.input("x", valid_type=Int)
        spec.input("y", valid_type=Int)
        spec.outline(
            cls.multiply,
            cls.result,
        )
        spec.output("result", valid_type=Int)

    def multiply(self):
        """Multiply two integers."""
        self.ctx.product = self.inputs.x * self.inputs.y

    def result(self):
        """Add the result to the outputs."""
        self.out("result", self.ctx.product)
```

3. Use a calcfunction
  * Add the `calcfuntion`-decorated `multiply` function. to the script
  * Use it inside the `multiply` step of the outline.

```python
@calcfunction
def multiply(x, y):
    return x * y

class MultiplyAddWorkChain(WorkChain):
    """WorkChain to multiply two numbers and add a third, for testing and demonstration purposes."""

    @classmethod
    def define(cls, spec):
        """Specify inputs and outputs."""
        super().define(spec)
        spec.input("x", valid_type=Int)
        spec.input("y", valid_type=Int)
        spec.outline(
            cls.multiply,
            cls.result,
        )
        spec.output("result", valid_type=Int)

    def multiply(self):
        """Multiply two integers."""
        self.ctx.product = multiply(self.inputs.x, self.inputs.y)

    def result(self):
        """Add the result to the outputs."""
        self.out("result", self.ctx.product)
```

4. Using a `CalcJob`
  * Adding `Code` input.
  * How to submit a `CalcJob` inside a work chain. They should know how to _run_ a `CalcJob` by now.
  * How to use the `ToContext` to pass the `CalcJob` node. (probably shouldn't use `future` here anymore, it's a little weird)
  * Obtain the outputs from the `CalcJob`. Note that the argument we passed to `ToContext` corresponds to the attribute of the `ctx` `AttributeDict`.

```python
ArithmeticAddCalculation = CalculationFactory("arithmetic.add")

@calcfunction
def multiply(x, y):
    return x * y

class MultiplyAddWorkChain(WorkChain):
    """WorkChain to multiply two numbers and add a third, for testing and demonstration purposes."""

    @classmethod
    def define(cls, spec):
        """Specify inputs and outputs."""
        super().define(spec)
        spec.input("x", valid_type=Int)
        spec.input("y", valid_type=Int)
        spec.input("z", valid_type=Int)
        spec.input("code", valid_type=Code)
        spec.outline(
            cls.multiply,
            cls.add,
            cls.result,
        )
        spec.output("result", valid_type=Int)

    def multiply(self):
        """Multiply two integers."""
        self.ctx.product = multiply(self.inputs.x, self.inputs.y)

    def add(self):
        """Add two numbers using the `ArithmeticAddCalculation` calculation job plugin."""
        inputs = {"x": self.ctx.product, "y": self.inputs.z, "code": self.inputs.code}
        future = self.submit(ArithmeticAddCalculation, **inputs)

        return ToContext(addition=future)

    def result(self):
        """Add the result to the outputs."""
        self.out("result", self.ctx.addition.outputs.sum)
```

And I think that's it! The error handling will be explained in the next module.

#### Seeing the work chain change.

I think it would be good that the participants execute the work chain in each step and see how it changes. We can use the RESTAPI/provenance graph or the `verdi process status` method. I think we should try this and experiment.

#### Debugging issues.

It would be ideal if the participants make these changes themselves. However, this means they will also have to be able to debug issues they encounter when running the work chain. This includes using the `verdi process status` method and `verdi process report`, and also checking the daemon logs? These are essential skills they need to know if they ever want to write workflows, and will actually help them learn by making mistakes.

#### `verdi daemon restart --reset`

This command can almost not appear too much in the tutorial. 🙃  We should also explain that the daemon needs to have access to the workflow they are writing by adjusting the `$PYTHON_PATH` in the `activate` script of the environment.

---

I think that'll do to kickstart writing this module. ;) The equation of state should not be in this module. It should really be all about the basics of writing a workflow as described above. Future modules will deal with:

* Error handling
* input validation
* non-linear outlines
* ...

@flavianojs don't hesitate to remove material if you think it doesn't fit in this module. I'll be keeping a close eye on removed content and see if I can find a home for it.

Thanks man! Let's make sure people can write top-notch workflows in just the one day we have available! I consider this to be the most critical part of the tutorial, but perhaps also the most challenging to write.

### PR DESCRIPTION
**Pull request by flavianojs**
We are attempting to construct a new Workflow section to be easier, step-by-step, and modular, see issue [#339 ](https://github.com/aiidateam/aiida-tutorials/issues/339).

### CODE REVIEW

**Comment by flavianojs** [comment]
@mbercx  and @ramirezfranciscof, please have a look at this first draft.

File: docs/sections/workflows/writing_basics.md [hunk 1]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+We used the `x` and `y` **inputs** of the work chain, as inputs for the calculation function.
+Note how we refer to the inputs of the work chain, e.g., `self.inputs.x`.
+Remember, `multiply()` should not (and cannot) generate data by itself.
+It has to rely on *calculation functions* like in the example, or on *calculation jobs*, which will be demonstrated later on.
+Then, there are other two instructions in `multiply()`:
+the first stores the `product` into the *context* of the work chain using `self.ctx` such that it can be used in the next step of the outline; and the second declares one of the outputs.
 
-```{literalinclude} include/code/multiply_add.py
+When we use the `self.out()`, the first argument (`'product'`) specifies the label of the output, which corresponds to the label provided to the spec in the `define()` method.
+The second argument is the result of the calc function.
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md]:
```suggestion
The second argument is the result of the calculation function, stored in the `product` Python variable.
```

File: docs/sections/workflows/writing_basics.md [hunk 2]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+
+        # Call `add` using a variable from the context and one of the inputs
+        result = add( self.ctx.product, self.inputs.z )
+
+        # Parsing the output
+        self.out("final_result", result)
+```
+
+In `add()`, we called the `addition()` calculation function (declared outside of the class) using a variable from the context and one of the inputs.
+Then, we declared the last output of the work chain.
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md]:
```suggestion
Finally, we once again call `self.out()` to link the node stored in the `result` variable as an output of the work chain with link label `"final_result"`.
```

File: docs/sections/workflows/writing_basics.md [hunk 3]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+And that is it.
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md]:
```suggestion
And that's it!
```

File: docs/sections/workflows/writing_basics.md [hunk 4]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+```{code-block} ipython
+In [1]: from aiida.engine import run
+In [2]: from multiply_add_workchain import MultiplyAddWorkChain
+In [3]: result = run(MultiplyAddWorkChain, Int(2), Int(3), Int(5) )
+In [4]: result
+Out[2]:
+{'product': <Int: uuid: 9c5c29de-6176-41fe-a051-672a5348e631 (pk: 1909) value: 6>,
+ 'final_result': <Int: uuid: 3dbebeb3-7ed4-4331-976c-94099a305622 (pk: 1911) value: 11>}
+```
+Check the list with last processes:
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md]: Maybe I'm overly repeating things, but I especially think reminding them of the `-p` option is a good idea:

```suggestion
Check the list of processes with `verdi process list`, using `-a / --all` to also see _terminated_ processes and `-p / --past-days 1` to only see processes _created_ in the past day:
```

File: docs/sections/workflows/writing_basics.md [hunk 5]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+------------  ----  ------
+final_result  1903  Int
+product       1901  Int
+
+Called      PK  Type
+--------  ----  --------------
+CALL      1900  multiplication
+CALL      1902  add
+```
+
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md:497]: Alright, this will definitely do for now for this section (the PR is already long, so tricky to keep focus on review). In *a different PR* (will open issue) we can still add more of a link between the labels they defined for the output links using the `self.out()` calls and what they see here.

File: docs/sections/workflows/writing_basics.md [hunk 6]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+:::{note}
+There are other ways to run/submit your work chain. ADD LINK.
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md]: Indeed good not to add here (tutorials shouldn't show multiple ways of doing things), but rather link to a how-to that explains this. 👍

File: docs/sections/workflows/writing_basics.md [hunk 7]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+:::
+
+#### Work chain with Calculation Jobs
+
+Here, we demonstrate how to include calculation jobs in our work chain.
+Calculation jobs represent external codes in AiiDA.
+We are going to use the example in Section {ref}`Constructing our first work chain<Constructing our first work chain>` and replace the addition calculation function:
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md]:
```suggestion
We are going to use the example in Section {ref}`Constructing our first work chain<Constructing our first work chain>` and replace the `addition` calculation function:
```
```suggestion
We are going to use the example in Section {ref}`Constructing our first work chain<Constructing our first work chain>` and replace the addition calculation function:
```

Besides that, these massive code blocks are hard to parse. I would just do:

... replace the `addition` calculation function:

_show **only** the old `add` step_

by the `ArithmeticAddCalculation` calculation _job_ instead:

_show **only** the new `add` step_

File: docs/sections/workflows/writing_basics.md [hunk 8]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+
+```{literalinclude} include/code/new_scripts/multiply_add_workchain_calcjob.py
+:language: python
+```
+
+Let us break down what we did.
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md:627]: If you agree with the above comment, this needs to be changed somewhat. But maybe we should first agree on that. ;)

File: docs/sections/workflows/writing_basics.md [hunk 9]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+In [2]: from multiply_add_workchain_calcjob import MultiplyAddWorkChain
+In [3]: add_code = load_code(label='add@localhost')
+In [4]: inputs = {'x': Int(1), 'y': Int(2), 'z': Int(3), 'code': add_code}
+In [5]: workchain_node = submit(MultiplyAddWorkChain, **inputs)
+```
+
+Now, go and check the last processes, the status of the `MultiplyAddWorkChain`, and show some details about the inputs and outputs.
+
+
+#### Checks and validation
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md:706]: As we've already discussed, this should be moved to a separate module. Note that another advantage of working with smaller pieces is that by now I've been reviewing too much material to still focus. Smaller PR's make for better reviews!

Comment (mbercx) [reply docs/sections/workflows/writing_basics.md:706]: I suggest I move and adapt this, and then you review :)

**Review by mbercx — CHANGES_REQUESTED** [review]
Thanks a **lot**, @flavianojs! This is already a _vast_ improvement compared to the previous version, and I think we're on the path to making a simple code-agnostic tutorial that will seriously lower the barrier for people to get started with writing their own workflows. Reading this, I realize that how you write workflows in AiiDA makes a lot of sense, and that it's _really not that difficult_.

I've left quite a bit of comments, but most of these are minor changes that you can simply fix using the "Add to batch feature" of GitHub. If you haven't done this yet, make sure you use it! Would be much more painful and error-prone to go and fix these manually (also I become a co-author 😏 ). You can decide which comments to add to batch and then commit them into one commit, that you can then `pull` into the local branch where you are developing. From there, you can continue to work on the other suggestions. Note that some things really _are just suggestions_. Better to quickly reply to a comment and discuss to than to make a bunch of changes first.

Some more GitHub-related comments:

* You can use [_closing keywords_](https://docs.github.com/en/issues/tracking-your-work-with-issues/creating-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) (like "fixes") to automatically link a PR to an issue. Just adding the issue hash-tag is not enough. Note that I now linked the PR to the issue manually.
* I see that your commits are not associated to your account. Have you [set your email address](https://docs.github.com/en/github/setting-up-and-managing-your-github-user-account/managing-email-preferences/setting-your-commit-email-address#:~:text=GitHub%20uses%20your%20commit%20email,commit%20email%20address%20on%20GitHub.) locally using `git`?
* We typically turn each PR into one commit on the `master` branch (`aiida-tutorials` has been a little different, but I want to move towards that except for quick small fixes). This means we'll [squash merge](https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/about-pull-request-merges) your PR in the end. The commit message is usually derived from the original post (OP) of the PR, so try to make that fully represents the changes in the PR.

Usually, I prefer that this final merging flow has the reviewer _approve_ the PR, and leaving the merging to the person who opened the PR since the commit message will be associated with their account. When you do merge the PR in the end, **be sure to select "Squash and merge". Then you can still edit the commit title and message.

**Comment by mbercx** [comment]
@flavianojs some additional, broader comments that I forgot to paste form my clipboard. 🙃  Happy to discuss these, after all, they are just suggestions, not _directives_.

Thanks again for your work! 🙏 

#### Executable exampes

I'm still not convinced the first example they execute has to be this large. As the [tutorial explanation page](https://documentation.divio.com/tutorials/#get-the-user-started) from the model that we follow says:

_"It’s also perfectly acceptable if what you get the beginner to do is not the way an experienced person would, or even if it’s not the ‘correct’ way - a tutorial for beginners is not the same thing as a manual for best practice."_

Just passing a node and outputting it is perhaps silly (even though still _instructive_, and they do something new), but I think having one step in the outline is fine. Not using the calc function at first will indeed fail, but maybe that's a good thing? It shows the concept that workflows cannot _create_ data.

I think part of the problem of the previous tutorial was they the weren't _doing_ enough that _teaches_ them something. In starting small and extending little by little - executing the work chain in every step - the user will see how the work chain changes due to the changes they make in the code. I think this is valuable.

#### Copy-pasta

I'm still a little worried that participants copying the content to their own module file will lead to all sorts of problems. Still considering if we should not just provide the modules in a directory that is pre-configured to be part of the daemon `PYTHON_PATH`. Copy pasting code isn't _doing_, in my opinion.

File: docs/sections/workflows/writing_basics.md [hunk 10]
```diff
@@ -44,93 +44,60 @@ In this section, you will learn to:
...
+7. Prepare input for code 3
+8. Run code 3
+9. Check the output from code 3
+10. Parse and save selected data
 
-For the simple `multiply` example, the output is guaranteed to be an `Int` node if the inputs are `Int` nodes.
-However, for more complex calculation functions you need to make sure that the function returns a `Data` node.
+Thus, in general, a careful scientist is doing many steps when performing a workflow by hand.
+Therefore, to automatize this process, we need to write a workflow that tries and mimics these various steps that we would do manually.
+Notice that instead of running a code, some steps could consist of workflows themselves.
```
[the remark this answers is not in the document]
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md]: Now that we want to split this first example is smaller steps, this info might come later down the line.

File: docs/sections/workflows/writing_basics.md [hunk 11]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
-Below is an example of a work chain that takes three integers as inputs, multiplies the first two and then adds the third to obtain the final result:
 
-```{literalinclude} include/code/multiply_add.py
-:language: python
-:start-after: start-marker
+#### Constructing our first work chain
 
-```
+Next, we are going to work, step-by-step, on an example of a work chain that takes three integers as inputs, multiplies the first two, and then adds the third to obtain the final result.
+Let get started by creating a file for our work chain (e.g. `multiply_add_workchain.py`), and adding the following piece of code:
```
[the remark this answers is not in the document]
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md]: Do you mean something beyond the diff bash command?

File: docs/sections/workflows/writing_basics.md [hunk 12]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
 
-You can give the work chain any valid Python class name, but the convention is to have it end in {class}`~aiida.engine.processes.workchains.workchain.WorkChain` so that it is always immediately clear what it references.
-Let's go over the methods of the `MultiplyAddWorkChain` one by one:
+```{code-block} python
```
[the remark this answers is not in the document]
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md]: The way I wrote the incremental changes to the workchain is similar to what we do when we are writing the file from scratch. Using snippets would be useful if we were just revealing pieces of the final code. Do you propose to use a `.py` file for each code block?

File: docs/sections/workflows/writing_basics.md [hunk 13]
```diff
@@ -175,103 +154,445 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
   Other keyword arguments allow the developer to set a default for the input, or indicate that an input should not be stored in the database, see {ref}`the process topics section <topics:processes:usage:spec>` for more details.
-* the **outline** or logic of the workflow, specified using the `spec.outline()` method.
+
+:::{note}
+
+All inputs and outputs of a work chain must be of AiiDA data types because only then they can be stored as Nodes in the AiiDA database.
+
+:::
+
+Next, let us add an outline to our work chain, i.e., a sequence of steps and the logic of the work chain (please modify your script along):
```
[the remark this answers is not in the document]
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md]: In my opinion, they should put their hands on it. Even better than copying it would be if they were to write it down following the instruction and example. I know you are worried about possible problems during the tutorial, but I think they are part of the process.

File: docs/sections/workflows/writing_basics.md (hunk 7 above)
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md]: I agree, I will try to implement this suggestion.

**Comment by mbercx** [comment]
@flavianojs let me know when this ready for another round of review! Also make sure to rebase on the `tutorial-2021-intro-week` branch (or merge the changes into this one, there should not be conflicts afaik). Let me know if you need a hand with this! :)

**Comment by mbercx** [comment]
Just to once again highlight the importance of making clear that the work chain module needs to be in the `PYTHONPATH` loaded by the daemon and this needs to be restarted before every submission: https://github.com/aiidateam/aiida-core/issues/4988

**Comment by flavianojs** [comment]
Hi @mbercx, I believe this PR is ready for the second round of review.

File: docs/sections/workflows/writing_basics.md [hunk 14]
```diff
@@ -175,103 +153,555 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+
+Thus, the information stored in the context can be used by another step of the outline.
+In our example, the `self.ctx.summation` was declarated as the output by the `result()` step:
+
+```{literalinclude} include/code/new_scripts/my_first_workchain_4_pass_context.py
+:language: python
+:pyobject: AddWorkChain.result
+```
+
+#### Two calculation functions and more outputs
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md]:
```suggestion
#### Adding more complexity
```

File: docs/sections/workflows/writing_basics.md [hunk 15]
```diff
@@ -175,103 +153,555 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
+
+Increasing the level of complexity, we want to write a work chain which receives three inputs, multiply the first two inputs and add to that result the value of the third input.
```
Comment (mbercx) [inline docs/sections/workflows/writing_basics.md:544]: Let's turn this section into an exercise! 😁

**Review by mbercx — CHANGES_REQUESTED** [review]
Alright, another quick pass at the section, mainly with linguistic comments. I suggest you make these small fixes and then we switch roles for the "adding complexity" exercise section and the "calculation job" one.

EDIT: Switch roles as in: I will adapt them in a new PR that you can review :)

File: docs/sections/workflows/writing_basics.md [hunk 16]
```diff
@@ -44,93 +44,60 @@ In this section, you will learn to:
...
+### Workflow in AiiDA
 
 A workflow in AiiDA is a {ref}`process <topics:processes:concepts>` that calls other workflows and calculations and optionally *returns* data and as such can encode the logic of a typical scientific workflow.
 Currently, there are two ways of implementing a workflow process:
 
 * {ref}`work functions<topics:workflows:concepts:workfunctions>`
 * {ref}`work chains<topics:workflows:concepts:workchains>`
 
-Here we present a brief introduction on how to write both workflow types.
+The main difference between them is that *work functions* are executed by the AiiDA daemons, while each step in a *work chain* generates a process of its own.
```
[the remark this answers is not in the document]
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md]: Just add a new sentece informing your item number 1.

File: docs/sections/workflows/writing_basics.md [hunk 17]
```diff
@@ -147,14 +114,25 @@ It is important to reiterate here that the {func}`~aiida.engine.processes.functi
...
-Copy the code snippet above and execute it in the `verdi shell`, or put it into a Python script (e.g. {download}`add_multiply.py <include/code/add_multiply.py>`) and import the add_multiply work function in the `verdi shell`:
+:::{note}
+
+In the above example, we created a workflow just to add then multiply two numbers.
+Please, keep in mind that in real situations instead of simple addition or multiplication, we would have complex calculations such as a DFT calculation or a data analysis.
+
+:::
+
+
+Copy the code snippet above and put it into a Python script (e.g. {download}`add_multiply.py <include/code/add_multiply.py>`).
```
[the remark this answers is not in the document]
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md:126]: I would not.

File: docs/sections/workflows/writing_basics.md [hunk 18]
```diff
@@ -175,103 +153,555 @@ See the {ref}`corresponding topics section for more details <topics:processes:us
...
-Next, the `define()` method should be used to define the specifications of the work chain, which are contained in the work chain `spec`:
+```{literalinclude} include/code/new_scripts/my_first_workchain_1_output_input.py
+:language: python
+:lines: 13
+```
 
-* the **inputs**, specified using the `spec.input()` method.
   The first argument of the `input()` method is a string that specifies the label of the input, e.g. `'x'`.
   The `valid_type` keyword argument allows you to specify the required node type of the input.
   Other keyword arguments allow the developer to set a default for the input, or indicate that an input should not be stored in the database, see {ref}`the process topics section <topics:processes:usage:spec>` for more details.
```
[the remark this answers is not in the document]
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md]: I agree.

File: docs/sections/workflows/writing_basics.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (mbercx) [reply docs/sections/workflows/writing_basics.md:126]: Yes, reading more through the module, and seeing we now also add an exercise, I agree :)

File: docs/sections/workflows/writing_basics.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (flavianojs) [reply docs/sections/workflows/writing_basics.md]: I did that. It will come up in the next commit

**Review by mbercx — APPROVED** [review]
All good! Thanks again for the great work, @flavianojs! 👍 🎉  I'll take care of the leftover comments after reorganising the content into modules.
